bpf/lib/bpf_cgroup.h

```diff
@@ -10,16 +10,91 @@
 
 #define NULL ((void *)0)
 
+#ifndef CGROUP_SUPER_MAGIC
+#define CGROUP_SUPER_MAGIC 0x27e0eb /* Cgroupv1 pseudo FS */
+#endif
+
+#ifndef CGROUP2_SUPER_MAGIC
+#define CGROUP2_SUPER_MAGIC 0x63677270 /* Cgroupv2 pseudo FS */
+#endif
+
 /* Our kernfs node name length, can be made 256? */
 #define KN_NAME_LENGTH 128
 
+/* Max nested cgroups that are tracked. Arbitrary value, nested cgroups
+ * that are at a level greater than 32 will be attached to the cgroup
+ * at level 32.
+ */
+#define CGROUP_MAX_NESTED_LEVEL 32
+
+typedef enum {
+	CGROUP_UNTRACKED = 0, /* Cgroup was created but we did not track it */
+	CGROUP_NEW = 1, /* Cgroup was just created */
+	CGROUP_RUNNING = 2, /* new => running (fork,exec task inside) */
+	CGROUP_RUNNING_PROC = 3, /* Generated from pids of procfs */
+} cgroup_state;
+
 /* Represent old kernfs node with the kernfs_node_id
  * union to read the id in 5.4 kernels and older
  */
 struct kernfs_node___old {
 	union kernfs_node_id id;
 };
 
+struct cgroup_tracking_value {
+	/* State of cgroup */
+	cgroup_state state;
+
+	/* Unique id for the hierarchy this is mostly for cgroupv1 */
+	__u32 hierarchy_id;
+
+	/* The depth this cgroup is at */
+	__u32 level;
+
+	__u32 pad;
+
+	/* Cgroup kernfs_node name */
+	char name[KN_NAME_LENGTH];
```

Comment:
I'm not sure keeping the string name is needed here. What would we use this for? I think ideally this would be in the event and then a cache in userspace would map the name to the id. Then BPF side mainly deals with just numeric ids and avoids strings altogether.

Reply:
I tend to agree, we made it this way in beginning so we can project how this could work from the upper layers when doing filtering, will recheck again with @kkourt and probably move it to user space.

Reply:
After discussing it offline with Djalal, it seems clear that we need a mapping from cgroup ids to names.  There does not seem to be a clear argument for keeping it in bpf and/or user-space. If we keep it in user-space, then we will avoid the syscall. If we keep it in bpf, it would persist agent restarts. Given that subsequent patches depend on having the data in the bpf map, I'd suggest we keep it as is. We can always rethink it later.

Reply:
Thanks @kkourt @jrfastab

Reply:
@jrfastab just an update here, I got confused we definitely need this at the bpf map since when we generate events (exec) we do get the tracking cgroup ID => lookup into the bpf map and find the tracking cgroup name potentially pod or container ID, and in next PR we will send this info part of the event: tracking cgroup ID + tracing cgroup Name